File: md_studio/storage/content_adapter.py
import os
import uuid
import aiofiles
from pathlib import Path
from datetime import datetime
from typing import Optional, List, Dict, Any, Literal
from dataclasses import dataclass, asdict
import frontmatter

from ..utils.excerpt import create_excerpt
from ..utils.safe_fs import ensure_dir, path_exists, atomic_write_file, atomic_write_json, safe_read_json
from ..utils.slug import slugify, ensure_unique_slug

SortBy = Literal["title-asc", "date-newest", "date-oldest", "updated-newest"]
FilterBy = Literal["all", "public", "private"]

@dataclass
class DocMeta:
    title: str
    slug: str
    excerpt: str
    isPublic: bool
    publicId: Optional[str]
    createdAt: str
    updatedAt: str

@dataclass
class DocFull:
    title: str
    slug: str
    excerpt: str
    isPublic: bool
    publicId: Optional[str]
    createdAt: str
    updatedAt: str
    bodyMd: str

class ContentAdapter:
    def __init__(self, content_root: Optional[str] = None, index_path: Optional[str] = None):
        self.content_root = Path(content_root or os.path.join(os.getcwd(), "content"))
        self.index_path = Path(index_path or os.path.join(os.getcwd(), ".md-studio", "index.json"))
        
        self.scan_roots = self._get_scan_roots()
        self.write_root = self._get_write_root()
    
    def _get_scan_roots(self) -> List[Path]:
        scan_dirs = os.getenv("SCAN_DIRS")
        if scan_dirs:
            return [Path(d.strip()) if os.path.isabs(d.strip()) else Path(os.getcwd()) / d.strip() 
                   for d in scan_dirs.split(",") if d.strip()]
        
        content_dirs = os.getenv("CONTENT_DIRS")
        if content_dirs:
            return [Path(d.strip()) if os.path.isabs(d.strip()) else Path(os.getcwd()) / d.strip() 
                   for d in content_dirs.split(",") if d.strip()]
        
        return [self.content_root]
    
    def _get_write_root(self) -> Path:
        write_dir = os.getenv("WRITE_DIR")
        if write_dir:
            dirs = [Path(d.strip()) if os.path.isabs(d.strip()) else Path(os.getcwd()) / d.strip() 
                   for d in write_dir.split(",") if d.strip()]
            if dirs:
                return dirs[0]
        return self.scan_roots[0] if self.scan_roots else self.content_root
    
    async def list(
        self,
        query: str = "",
        page: int = 1,
        page_size: int = 20,
        sort_by: SortBy = "date-newest",
        filter_by: FilterBy = "all"
    ) -> Dict[str, Any]:
        index = await self._read_index()
        normalized_query = query.strip().lower()
        
        if filter_by == "public":
            index = [entry for entry in index if entry.isPublic]
        elif filter_by == "private":
            index = [entry for entry in index if not entry.isPublic]
        
        filtered = index
        if normalized_query:
            filtered = [
                entry for entry in index
                if normalized_query in f"{entry.title} {entry.slug} {entry.excerpt}".lower()
            ]
        
        sorted_items = sorted(filtered, key=lambda x: self._sort_key(x, sort_by))
        
        total = len(sorted_items)
        start = max(0, (page - 1) * page_size)
        end = start + page_size
        items = sorted_items[start:end]
        
        return {
            "items": [asdict(item) for item in items],
            "total": total
        }
    
    def _sort_key(self, item: DocMeta, sort_by: SortBy):
        if sort_by == "title-asc":
            return item.title.lower()
        elif sort_by == "date-oldest":
            return datetime.fromisoformat(item.createdAt)
        elif sort_by == "updated-newest":
            return -datetime.fromisoformat(item.updatedAt).timestamp()
        else:
            return -datetime.fromisoformat(item.createdAt).timestamp()
    
    async def get_by_slug(self, slug: str) -> Optional[Dict[str, Any]]:
        location = await self._find_doc_location(slug)
        if not location:
            return None
        
        async with aiofiles.open(location["path"], 'r', encoding='utf-8') as f:
            content = await f.read()
        
        post = frontmatter.loads(content)
        data = self._normalize_frontmatter(post.metadata, slug)
        
        body_md = post.content.rstrip()
        meta_data = dict(data)
        excerpt = meta_data.pop("excerpt", "") or create_excerpt(body_md)
        meta = DocMeta(
            **meta_data,
            excerpt=excerpt,
        )
        
        result = asdict(meta)
        result["bodyMd"] = body_md
        return result
    
    async def get_by_public_id(self, public_id: str) -> Optional[Dict[str, Any]]:
        index = await self._read_index()
        match = next((entry for entry in index if entry.publicId == public_id), None)
        if not match:
            return None
        
        return await self.get_by_slug(match.slug)
    
    async def create(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        title = input_data["title"].strip()
        body_md = input_data["bodyMd"]
        desired_slug = input_data.get("slug")
        
        slug = await ensure_unique_slug(
            desired_slug,
            title,
            lambda candidate: self._is_slug_available(candidate)
        )
        
        if not slug:
            raise ValueError("Unable to derive slug from title.")
        
        now = datetime.now().isoformat()
        doc_meta = DocMeta(
            title=title,
            slug=slug,
            excerpt=create_excerpt(body_md),
            isPublic=False,
            publicId=None,
            createdAt=now,
            updatedAt=now
        )
        
        await self._write_document_file(self.write_root, doc_meta, body_md)
        await self._persist_index(lambda records: records + [doc_meta])
        
        return asdict(doc_meta)
    
    async def update(self, slug: str, patch: Dict[str, Any]) -> Dict[str, Any]:
        location = await self._find_doc_location(slug)
        if not location:
            raise ValueError(f'Document "{slug}" not found.')
        
        index = await self._read_index()
        target_index = next((i for i, entry in enumerate(index) if entry.slug == slug), -1)
        if target_index == -1:
            raise ValueError(f'Document "{slug}" not found.')
        
        current_meta = index[target_index]
        existing = await self.get_by_slug(slug)
        if not existing:
            raise ValueError(f'Document "{slug}" content missing.')
        
        desired_slug = slugify(patch.get("slug", "")) if patch.get("slug") else current_meta.slug
        if not desired_slug:
            raise ValueError("Updated slug cannot be empty.")
        
        next_slug = desired_slug
        if patch.get("slug") and desired_slug != slug:
            next_slug = await ensure_unique_slug(
                desired_slug,
                current_meta.title,
                lambda candidate: self._is_slug_available(candidate, exclude=slug)
            )
        
        now = datetime.now().isoformat()
        body_md = patch.get("bodyMd", existing["bodyMd"])
        
        is_public = patch.get("isPublic", current_meta.isPublic)
        public_id = current_meta.publicId
        if patch.get("isPublic") is True and not public_id:
            public_id = str(uuid.uuid4())
        elif patch.get("isPublic") is False:
            public_id = None
        
        updated_meta = DocMeta(
            title=patch.get("title", current_meta.title).strip() if patch.get("title") else current_meta.title,
            slug=next_slug,
            isPublic=is_public,
            publicId=public_id,
            excerpt=create_excerpt(body_md),
            createdAt=current_meta.createdAt,
            updatedAt=now
        )
        
        await self._write_document_file(Path(location["root"]), updated_meta, body_md)
        
        if next_slug != slug:
            old_path = self._get_doc_path(Path(location["root"]), slug)
            if await path_exists(old_path):
                os.remove(old_path)
        
        index[target_index] = updated_meta
        await self._write_index(index)
        
        return asdict(updated_meta)
    
    async def remove(self, slug: str) -> None:
        index = await self._read_index()
        next_index = [entry for entry in index if entry.slug != slug]
        if len(next_index) == len(index):
            raise ValueError(f'Document "{slug}" not found.')
        
        location = await self._find_doc_location(slug)
        if location:
            doc_path = self._get_doc_path(Path(location["root"]), slug)
            if await path_exists(doc_path):
                os.remove(doc_path)
        
        await self._write_index(next_index)
    
    async def toggle_public(self, slug: str) -> Dict[str, Any]:
        location = await self._find_doc_location(slug)
        if not location:
            raise ValueError(f'Document "{slug}" not found.')
        
        index = await self._read_index()
        target_index = next((i for i, entry in enumerate(index) if entry.slug == slug), -1)
        if target_index == -1:
            raise ValueError(f'Document "{slug}" not found.')
        
        meta = index[target_index]
        now_public = not meta.isPublic
        
        updated_meta = DocMeta(
            title=meta.title,
            slug=meta.slug,
            excerpt=meta.excerpt,
            isPublic=now_public,
            publicId=meta.publicId or str(uuid.uuid4()) if now_public else None,
            createdAt=meta.createdAt,
            updatedAt=datetime.now().isoformat()
        )
        
        document = await self.get_by_slug(slug)
        if not document:
            raise ValueError(f'Document "{slug}" content missing.')
        
        await self._write_document_file(Path(location["root"]), updated_meta, document["bodyMd"])
        index[target_index] = updated_meta
        await self._write_index(index)
        
        return asdict(updated_meta)
    
    async def export_raw(self, slug: str) -> Dict[str, str]:
        location = await self._find_doc_location(slug)
        if not location:
            raise ValueError(f'Document "{slug}" not found.')
        
        async with aiofiles.open(location["path"], 'r', encoding='utf-8') as f:
            content = await f.read()
        
        return {"filename": f"{slug}.md", "content": content}
    
    def _get_doc_path(self, root: Path, slug: str) -> Path:
        return root / f"{slug}.md"
    
    async def _find_doc_location(self, slug: str) -> Optional[Dict[str, str]]:
        for root in self.scan_roots:
            file_path = self._get_doc_path(root, slug)
            if await path_exists(file_path):
                return {"root": str(root), "path": str(file_path)}
        return None
    
    async def _is_slug_available(self, slug: str, exclude: Optional[str] = None) -> bool:
        if exclude and slug == exclude:
            return True
        
        for root in self.scan_roots:
            if await path_exists(self._get_doc_path(root, slug)):
                return False
        return True
    
    async def _read_index(self) -> List[DocMeta]:
        entries_data = await safe_read_json(self.index_path, [])
        entries = [DocMeta(**entry) for entry in entries_data]
        if entries:
            entries = self._merge_indexes([entries])
        
        index_exists = await path_exists(self.index_path)
        if not index_exists or not entries:
            disk_entries = await self._read_docs_from_scan_roots()
            if disk_entries:
                await self._write_index(disk_entries)
                return disk_entries
        
        return sorted(entries, key=lambda x: -datetime.fromisoformat(x.updatedAt).timestamp())
    
    async def _write_index(self, entries: List[DocMeta]) -> None:
        deduped = self._merge_indexes([entries]) if entries else []
        sorted_entries = sorted(deduped, key=lambda x: -datetime.fromisoformat(x.updatedAt).timestamp())
        await atomic_write_json(self.index_path, [asdict(e) for e in sorted_entries])
    
    async def _persist_index(self, mutator) -> None:
        current = await self._read_index()
        next_entries = mutator(current)
        if next_entries:
            next_entries = self._merge_indexes([next_entries])
        await self._write_index(next_entries)
    
    def _normalize_frontmatter(self, data: Dict[str, Any], slug: str, stats: Optional[Dict] = None) -> Dict[str, Any]:
        fallback_created = stats.get("birthtime", datetime.now()) if stats else datetime.now()
        fallback_updated = stats.get("mtime", datetime.now()) if stats else datetime.now()
        
        return {
            "title": data.get("title", slug),
            "slug": data.get("slug", slug),
            "excerpt": data.get("excerpt", ""),
            "isPublic": bool(data.get("isPublic", False)),
            "publicId": data.get("publicId") if data.get("publicId") else None,
            "createdAt": data.get("createdAt", fallback_created.isoformat() if isinstance(fallback_created, datetime) else fallback_created),
            "updatedAt": data.get("updatedAt", fallback_updated.isoformat() if isinstance(fallback_updated, datetime) else fallback_updated)
        }
    
    async def _write_document_file(self, root: Path, meta: DocMeta, body_md: str) -> None:
        metadata = {
            "title": meta.title,
            "slug": meta.slug,
            "isPublic": meta.isPublic,
            "publicId": meta.publicId,
            "createdAt": meta.createdAt,
            "updatedAt": meta.updatedAt
        }
        
        post = frontmatter.Post(body_md.rstrip() + "\n", **metadata)
        markdown_content = frontmatter.dumps(post)
        
        await ensure_dir(root)
        await atomic_write_file(self._get_doc_path(root, meta.slug), markdown_content)
    
    async def _read_docs_from_disk(self, root: Path) -> List[DocMeta]:
        if not await path_exists(root):
            return []
        
        try:
            entries = os.listdir(root)
        except Exception:
            return []
        
        markdown_files = [entry for entry in entries if entry.endswith(".md")]
        
        docs = []
        for filename in markdown_files:
            slug = filename.replace(".md", "")
            file_path = root / filename
            
            async with aiofiles.open(file_path, 'r', encoding='utf-8') as f:
                content = await f.read()
            
            post = frontmatter.loads(content)
            stats = os.stat(file_path)
            data = self._normalize_frontmatter(
                post.metadata, 
                slug,
                {"birthtime": datetime.fromtimestamp(stats.st_ctime), "mtime": datetime.fromtimestamp(stats.st_mtime)}
            )
            
            body_content = post.content.rstrip()
            meta_data = dict(data)
            excerpt = meta_data.pop("excerpt", "") or create_excerpt(body_content)
            doc_meta = DocMeta(
                **meta_data,
                excerpt=excerpt,
            )
            docs.append(doc_meta)
        
        return docs
    
    async def _read_docs_from_scan_roots(self) -> List[DocMeta]:
        all_docs = []
        for root in self.scan_roots:
            docs = await self._read_docs_from_disk(root)
            all_docs.append(docs)
        
        return self._merge_indexes(all_docs)
    
    def _merge_indexes(self, indexes: List[List[DocMeta]]) -> List[DocMeta]:
        by_slug: Dict[str, DocMeta] = {}
        for entries in indexes:
            for entry in entries:
                existing = by_slug.get(entry.slug)
                if not existing:
                    by_slug[entry.slug] = entry
                    continue
                
                existing_time = datetime.fromisoformat(existing.updatedAt).timestamp()
                entry_time = datetime.fromisoformat(entry.updatedAt).timestamp()
                if entry_time >= existing_time:
                    by_slug[entry.slug] = entry
        
        return list(by_slug.values())

File: ui/lib/content-adapter.ts
import { randomUUID } from "crypto";
import { promises as fs, watch as fsWatch } from "fs";
import matter from "gray-matter";
import * as path from "path";

import { createExcerpt } from "@/lib/excerpt";
import {
  atomicWriteFile,
  atomicWriteJson,
  ensureDir,
  pathExists,
  safeReadJson,
} from "@/lib/safe-fs";
import { ensureUniqueSlug, slugify } from "@/lib/slug";

export type DocMeta = {
  title: string;
  slug: string;
  excerpt: string;
  isPublic: boolean;
  publicId: string | null;
  createdAt: string;
  updatedAt: string;
};

export type DocFull = DocMeta & {
  bodyMd: string;
};

export type SortBy = "title-asc" | "date-newest" | "date-oldest" | "updated-newest";
export type FilterBy = "all" | "public" | "private";

export interface ContentAdapter {
  list(
    query?: string,
    page?: number,
    pageSize?: number,
    sortBy?: SortBy,
    filterBy?: FilterBy,
  ): Promise<{ items: DocMeta[]; total: number }>;
  getBySlug(slug: string): Promise<DocFull | null>;
  getByPublicId(publicId: string): Promise<DocFull | null>;
  create(input: {
    title: string;
    slug?: string;
    bodyMd: string;
  }): Promise<DocMeta>;
  update(
    slug: string,
    patch: {
      title?: string;
      slug?: string;
      bodyMd?: string;
      isPublic?: boolean;
    },
  ): Promise<DocMeta>;
  remove(slug: string): Promise<void>;
  togglePublic(slug: string): Promise<DocMeta>;
  exportRaw(slug: string): Promise<{ filename: string; content: string }>;
}

const DEFAULT_CONTENT_ROOT = path.join(process.cwd(), "content");
const INDEX_PATH = path.join(process.cwd(), ".md-studio", "index.json");
const WATCH_DEBOUNCE_MS = 400;
let watcherInitialized = false;
let watcherTimer: NodeJS.Timeout | null = null;
let pollTimer: NodeJS.Timeout | null = null;
const activeWatchers: ReturnType<typeof fsWatch>[] = [];

function shouldEnableWatcher(): boolean {
  const flag = (process.env.CONTENT_WATCHER ?? "").toLowerCase();
  if (flag === "false" || flag === "0") {
    return false;
  }
  const mode = (process.env.STORAGE_ADAPTER ?? "fs").toLowerCase();
  return mode === "fs";
}

function getPollIntervalMs(): number {
  const raw = process.env.CONTENT_WATCH_POLL_MS;
  if (!raw) {
    return 0;
  }
  const parsed = Number(raw);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : 0;
}

function normalizeRoots(raw: string | undefined): string[] {
  if (!raw) {
    return [];
  }

  return raw
    .split(",")
    .map((entry) => entry.trim())
    .filter(Boolean)
    .map((entry) => (path.isAbsolute(entry) ? entry : path.join(process.cwd(), entry)));
}

function getScanRoots(): string[] {
  const scanRoots = normalizeRoots(process.env.SCAN_DIRS);
  if (scanRoots.length) {
    return scanRoots;
  }

  const legacyRoots = normalizeRoots(process.env.CONTENT_DIRS);
  if (legacyRoots.length) {
    return legacyRoots;
  }

  return [DEFAULT_CONTENT_ROOT];
}

function getWriteRoot(scanRoots: string[]): string {
  const explicit = normalizeRoots(process.env.WRITE_DIR);
  if (explicit.length) {
    return explicit[0];
  }
  return scanRoots[0] ?? DEFAULT_CONTENT_ROOT;
}

class LocalFsAdapter implements ContentAdapter {
  constructor() {
    if (!watcherInitialized && shouldEnableWatcher()) {
      watcherInitialized = true;
      void this.startWatcher();
    }
  }

  async list(
    query?: string,
    page = 1,
    pageSize = 20,
    sortBy: SortBy = "date-newest",
    filterBy: FilterBy = "all",
  ) {
    let index = await this.readIndex();
    const normalizedQuery = query?.trim().toLowerCase() ?? "";

    // Apply filter
    if (filterBy === "public") {
      index = index.filter((entry) => entry.isPublic);
    } else if (filterBy === "private") {
      index = index.filter((entry) => !entry.isPublic);
    }

    // Apply search
    const filtered = normalizedQuery
      ? index.filter((entry) => {
          const haystack = `${entry.title} ${entry.slug} ${entry.excerpt}`.toLowerCase();
          return haystack.includes(normalizedQuery);
        })
      : index;

    // Apply sorting
    const sorted = [...filtered].sort((a, b) => {
      switch (sortBy) {
        case "title-asc":
          return a.title.localeCompare(b.title);
        case "date-oldest":
          return new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime();
        case "updated-newest":
          return new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime();
        case "date-newest":
        default:
          return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime();
      }
    });

    const total = sorted.length;
    const start = Math.max(0, (page - 1) * pageSize);
    const end = start + pageSize;
    const items = sorted.slice(start, end);

    return { items, total };
  }

  async getBySlug(slug: string) {
    const location = await this.findDocLocation(slug);
    if (!location) {
      return null;
    }

    const raw = await fs.readFile(location.path, "utf8");
    const parsed = matter(raw);
    const data = this.normalizeFrontMatter(parsed.data, slug);

    const content = parsed.content.trimEnd();
    const meta = {
      ...data,
      excerpt: data.excerpt || createExcerpt(content),
    };

    return {
      ...meta,
      bodyMd: content,
    };
  }

  async getByPublicId(publicId: string) {
    const index = await this.readIndex();
    const match = index.find((entry) => entry.publicId === publicId);
    if (!match) {
      return null;
    }

    return this.getBySlug(match.slug);
  }

  async create(input: { title: string; slug?: string; bodyMd: string }) {
    const roots = this.getRoots();
    const primaryRoot = getWriteRoot(roots);
    const slug = await ensureUniqueSlug(
      input.slug,
      input.title,
      async (candidate) => {
        const existing = await Promise.all(
          roots.map((root) => pathExists(this.getDocPath(root, candidate))),
        );
        return !existing.some(Boolean);
      },
    );
    if (!slug) {
      throw new Error("Unable to derive slug from title.");
    }

    const now = new Date().toISOString();
    const docMeta: DocMeta = {
      title: input.title.trim(),
      slug,
      excerpt: createExcerpt(input.bodyMd),
      isPublic: false,
      publicId: null,
      createdAt: now,
      updatedAt: now,
    };

    await this.writeDocumentFile(primaryRoot, docMeta, input.bodyMd);
    await this.persistIndex((records) => [...records, docMeta]);

    return docMeta;
  }

  async update(
    slug: string,
    patch: {
      title?: string;
      slug?: string;
      bodyMd?: string;
      isPublic?: boolean;
    },
  ) {
    const location = await this.findDocLocation(slug);
    if (!location) {
      throw new Error(`Document "${slug}" not found.`);
    }

    const index = await this.readIndex();
    const targetIndex = index.findIndex((entry) => entry.slug === slug);
    if (targetIndex === -1) {
      throw new Error(`Document "${slug}" not found.`);
    }

    const currentMeta = index[targetIndex];
    const existing = await this.getBySlug(slug);
    if (!existing) {
      throw new Error(`Document "${slug}" content missing.`);
    }

    const desiredSlug = patch.slug ? slugify(patch.slug) : currentMeta.slug;
    if (!desiredSlug) {
      throw new Error("Updated slug cannot be empty.");
    }

    let nextSlug = desiredSlug;
    if (patch.slug && desiredSlug !== slug) {
      const roots = this.getRoots();
      nextSlug = await ensureUniqueSlug(
        desiredSlug,
        currentMeta.title,
        async (candidate) => {
          if (candidate === slug) {
            return true;
          }
          const existing = await Promise.all(
            roots.map((root) => pathExists(this.getDocPath(root, candidate))),
          );
          return !existing.some(Boolean);
        },
      );
    }

    const now = new Date().toISOString();
    const bodyMd = patch.bodyMd ?? existing.bodyMd;
    const updatedMeta: DocMeta = {
      ...currentMeta,
      title: patch.title?.trim() ?? currentMeta.title,
      slug: nextSlug,
      isPublic: patch.isPublic ?? currentMeta.isPublic,
      publicId:
        patch.isPublic === false ? null : currentMeta.publicId ?? existing.publicId,
      excerpt: createExcerpt(bodyMd),
      updatedAt: now,
    };

    if (patch.isPublic === true && !updatedMeta.publicId) {
      updatedMeta.publicId = currentMeta.publicId ?? randomUUID();
    }

    await this.writeDocumentFile(location.root, updatedMeta, bodyMd);

    if (nextSlug !== slug) {
      await fs.rm(this.getDocPath(location.root, slug), { force: true });
    }

    index[targetIndex] = updatedMeta;
    await this.writeIndex(index);

    return updatedMeta;
  }

  async remove(slug: string) {
    const index = await this.readIndex();
    const nextIndex = index.filter((entry) => entry.slug !== slug);
    if (nextIndex.length === index.length) {
      throw new Error(`Document "${slug}" not found.`);
    }

    const location = await this.findDocLocation(slug);
    if (location) {
      await fs.rm(this.getDocPath(location.root, slug), { force: true });
    }
    await this.writeIndex(nextIndex);
  }

  async togglePublic(slug: string) {
    const location = await this.findDocLocation(slug);
    if (!location) {
      throw new Error(`Document "${slug}" not found.`);
    }

    const index = await this.readIndex();
    const targetIndex = index.findIndex((entry) => entry.slug === slug);
    if (targetIndex === -1) {
      throw new Error(`Document "${slug}" not found.`);
    }

    const meta = index[targetIndex];
    const nowPublic = !meta.isPublic;
    const updatedMeta: DocMeta = {
      ...meta,
      isPublic: nowPublic,
      publicId: nowPublic ? meta.publicId ?? randomUUID() : null,
      updatedAt: new Date().toISOString(),
    };

    const document = await this.getBySlug(slug);
    if (!document) {
      throw new Error(`Document "${slug}" content missing.`);
    }

    await this.writeDocumentFile(location.root, updatedMeta, document.bodyMd);
    index[targetIndex] = updatedMeta;
    await this.writeIndex(index);

    return updatedMeta;
  }

  async exportRaw(slug: string) {
    const location = await this.findDocLocation(slug);
    if (!location) {
      throw new Error(`Document "${slug}" not found.`);
    }

    const content = await fs.readFile(location.path, "utf8");
    return { filename: `${slug}.md`, content };
  }

  private getDocPath(root: string, slug: string) {
    return path.join(root, `${slug}.md`);
  }
  private getRoots() {
    const scanRoots = getScanRoots();
    const writeRoot = getWriteRoot(scanRoots);
    const roots = new Set<string>(scanRoots);
    roots.add(writeRoot);
    return Array.from(roots);
  }

  private async findDocLocation(
    slug: string,
  ): Promise<{ root: string; path: string } | null> {
    const roots = this.getRoots();
    for (const root of roots) {
      const filePath = this.getDocPath(root, slug);
      if (await pathExists(filePath)) {
        return { root, path: filePath };
      }
    }
    return null;
  }

  private async readIndex(): Promise<DocMeta[]> {
    const entries = await safeReadJson<DocMeta[]>(INDEX_PATH, []);
    const indexExists = await pathExists(INDEX_PATH);
    if (!indexExists || entries.length === 0) {
      const diskEntries = await this.readDocsFromScanRoots();
      if (diskEntries.length > 0) {
        await this.writeIndex(diskEntries);
        return diskEntries;
      }
    }
    const deduped = this.mergeIndexes([entries]);
    if (deduped.length !== entries.length) {
      await this.writeIndex(deduped);
      return deduped;
    }
    return entries.sort(
      (a, b) =>
        new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime(),
    );
  }

  private async writeIndex(entries: DocMeta[]) {
    const sorted = [...entries].sort(
      (a, b) =>
        new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime(),
    );
    await ensureDir(path.dirname(INDEX_PATH));
    await atomicWriteJson(INDEX_PATH, sorted);
  }

  private async persistIndex(
    mutator: (records: DocMeta[]) => DocMeta[],
  ): Promise<void> {
    const current = await this.readIndex();
    const next = mutator(current);
    await this.writeIndex(next);
  }

  private normalizeFrontMatter(
    data: Record<string, unknown>,
    slug: string,
    stats?: { birthtime: Date; mtime: Date },
  ): DocMeta {
    const fallbackCreatedAt = stats?.birthtime ?? new Date();
    const fallbackUpdatedAt = stats?.mtime ?? new Date();
    return {
      title: typeof data.title === "string" ? data.title : slug,
      slug: typeof data.slug === "string" ? data.slug : slug,
      excerpt: typeof data.excerpt === "string" ? data.excerpt : "",
      isPublic: Boolean(data.isPublic),
      publicId:
        typeof data.publicId === "string" && data.publicId.length > 0
          ? data.publicId
          : null,
      createdAt:
        typeof data.createdAt === "string"
          ? data.createdAt
          : fallbackCreatedAt.toISOString(),
      updatedAt:
        typeof data.updatedAt === "string"
          ? data.updatedAt
          : fallbackUpdatedAt.toISOString(),
    };
  }

  private async writeDocumentFile(root: string, meta: DocMeta, bodyMd: string) {
    const frontMatter = {
      title: meta.title,
      slug: meta.slug,
      isPublic: meta.isPublic,
      publicId: meta.publicId,
      createdAt: meta.createdAt,
      updatedAt: meta.updatedAt,
    };

    const markdown = matter.stringify(bodyMd.trimEnd() + "\n", frontMatter);
    await ensureDir(root);
    await atomicWriteFile(this.getDocPath(root, meta.slug), markdown);
  }

  private async readDocsFromDisk(root: string): Promise<DocMeta[]> {
    if (!(await pathExists(root))) {
      return [];
    }

    let entries: string[];
    try {
      entries = await fs.readdir(root);
    } catch {
      return [];
    }
    const markdownFiles = entries.filter((entry) => entry.endsWith(".md"));

    const docs = await Promise.all(
      markdownFiles.map(async (filename) => {
        const slug = filename.replace(/\.md$/, "");
        const filePath = path.join(root, filename);
        const raw = await fs.readFile(filePath, "utf8");
        const parsed = matter(raw);
        const stats = await fs.stat(filePath);
        const data = this.normalizeFrontMatter(parsed.data, slug, stats);
        const content = parsed.content.trimEnd();
        return {
          ...data,
          excerpt: data.excerpt || createExcerpt(content),
        };
      }),
    );

    return docs;
  }

  private async readDocsFromScanRoots(): Promise<DocMeta[]> {
    const roots = this.getRoots();
    const entries = await Promise.all(
      roots.map((root) => this.readDocsFromDisk(root)),
    );
    return this.mergeIndexes(entries);
  }

  private mergeIndexes(indexes: DocMeta[][]): DocMeta[] {
    const bySlug = new Map<string, DocMeta>();
    for (const entries of indexes) {
      for (const entry of entries) {
        const existing = bySlug.get(entry.slug);
        if (!existing) {
          bySlug.set(entry.slug, entry);
          continue;
        }
        const existingTime = new Date(existing.updatedAt).getTime();
        const entryTime = new Date(entry.updatedAt).getTime();
        if (entryTime >= existingTime) {
          bySlug.set(entry.slug, entry);
        }
      }
    }
    return Array.from(bySlug.values());
  }

  private indexesEqual(a: DocMeta[], b: DocMeta[]): boolean {
    if (a.length !== b.length) {
      return false;
    }
    const normalize = (entries: DocMeta[]) =>
      [...entries].sort((left, right) => left.slug.localeCompare(right.slug));
    const aSorted = normalize(a);
    const bSorted = normalize(b);
    for (let i = 0; i < aSorted.length; i += 1) {
      const left = aSorted[i];
      const right = bSorted[i];
      if (left.slug !== right.slug) {
        return false;
      }
      if (
        left.title !== right.title ||
        left.excerpt !== right.excerpt ||
        left.isPublic !== right.isPublic ||
        left.publicId !== right.publicId ||
        left.createdAt !== right.createdAt ||
        left.updatedAt !== right.updatedAt
      ) {
        return false;
      }
    }
    return true;
  }

  private scheduleRescan(reason: string) {
    if (watcherTimer) {
      clearTimeout(watcherTimer);
    }
    watcherTimer = setTimeout(() => {
      void this.syncIndexWithDisk(reason);
    }, WATCH_DEBOUNCE_MS);
  }

  private async syncIndexWithDisk(reason: string) {
    try {
      const diskEntries = await this.readDocsFromScanRoots();
      const current = await safeReadJson<DocMeta[]>(INDEX_PATH, []);
      if (!this.indexesEqual(current, diskEntries)) {
        await this.writeIndex(diskEntries);
      }
    } catch (error) {
      console.warn(`Content watcher resync failed (${reason}).`, error);
    }
  }

  private async startWatcher() {
    const scanRoots = getScanRoots();
    const writeRoot = getWriteRoot(scanRoots);
    const roots = this.getRoots();

    await ensureDir(writeRoot);
    for (const root of roots) {
      if (!(await pathExists(root))) {
        continue;
      }
      try {
        const watcher = fsWatch(root, { persistent: true }, (event, filename) => {
          if (filename && !filename.toString().endsWith(".md")) {
            return;
          }
          this.scheduleRescan(`${event}:${filename ?? "unknown"}`);
        });
        activeWatchers.push(watcher);
      } catch (error) {
        console.warn(`Content watcher failed to start for ${root}.`, error);
      }
    }

    const pollInterval = getPollIntervalMs();
    if (pollInterval > 0) {
      pollTimer = setInterval(() => {
        this.scheduleRescan("poll");
      }, pollInterval);
    }

    await this.syncIndexWithDisk("startup");
  }
}

class S3OrGitAdapter implements ContentAdapter {
  constructor(private readonly mode: "s3" | "git") {}

  private notImplemented(): never {
    throw new Error(
      `The ${this.mode.toUpperCase()} content adapter is not implemented yet. ` +
        "See README for guidance on enabling S3 or Git-based storage.",
    );
  }

  async list(): Promise<{ items: DocMeta[]; total: number }> {
    this.notImplemented();
  }

  async getBySlug(): Promise<DocFull | null> {
    this.notImplemented();
  }

  async getByPublicId(): Promise<DocFull | null> {
    this.notImplemented();
  }

  async create(): Promise<DocMeta> {
    this.notImplemented();
  }

  async update(): Promise<DocMeta> {
    this.notImplemented();
  }

  async remove(): Promise<void> {
    this.notImplemented();
  }

  async togglePublic(): Promise<DocMeta> {
    this.notImplemented();
  }

  async exportRaw(): Promise<{ filename: string; content: string }> {
    this.notImplemented();
  }
}

let cachedAdapter: ContentAdapter | null = null;

export function getContentAdapter(): ContentAdapter {
  if (cachedAdapter) {
    return cachedAdapter;
  }

  const mode = (process.env.STORAGE_ADAPTER ?? "fs").toLowerCase();

  if (mode === "s3" || mode === "git") {
    cachedAdapter = new S3OrGitAdapter(mode);
  } else {
    cachedAdapter = new LocalFsAdapter();
  }

  return cachedAdapter;
}

export { LocalFsAdapter, S3OrGitAdapter };
